### ISSUE TITLE
Audio Unit Format (AU,.component）for macOS please

### ISSUE THREAD

**Issue by Darling-Lee** [michaelwillis/dragonfly-reverb#126]
Audio Unit Format (AU,.component）for macOS please

**Comment by michaelwillis** [comment]
That will depend upon [Distrho Plugin Framework](https://github.com/DISTRHO/DPF) supporting AU components on MacOS. See https://github.com/DISTRHO/DPF/issues/16

**Comment by falkTX** [comment]
While DPF does not support AU yet, one possible alternative is leveraging juce+dpf together just to do AU builds.
I have this setup for Cardinal at https://github.com/DISTRHO/Cardinal/tree/main/jucewrapper

Before this though, the plugin needs to handle high-dpi. Right now it simply ignored the scale factor and draws at 1x.
Some rework of the code is needed in order to make it look nice and crisp on high-dpi.

**Comment by falkTX** [comment]
I mean in order to supper high-dpi, each individual plugin needs specific code to support it.
DPF has the means to support this internally with an automatic upscaling, it will look blurry but have the correct size.

That auto-scaling is done through [setGeometryConstraints](https://distrho.github.io/DPF/classWindow.html#aa3824e300206b15241df8c8c78785d73) by setting `automaticallyScale` to true.
Then you can test with `export DPF_SCALE_FACTOR=2` to force scaling to 2x.

Alternatively, use `getScaleFactor()` to know the factor by which to scale up the UI. And do it all manually yourself.

**Comment by michaelwillis** [comment]
@falkTX I tried following the example of your jucewrapper from Cardinal. I think I almost got it working, but the linker fails:

```
Undefined symbols for architecture arm64:
  "DISTRHO::d_nextBufferSize", referenced from:
      JuceAU::JuceAU(ComponentInstanceRecord*) in lto.o
  "DISTRHO::g_nextBundlePath", referenced from:
      DISTRHO::DragonflyWrapperEditor::paint(juce::Graphics&) in lto.o
  "DISTRHO::UI::PrivateData::s_nextPrivateData", referenced from:
      DISTRHO::DragonflyWrapperEditor::paint(juce::Graphics&) in lto.o
  "DISTRHO::createUI()", referenced from:
      DISTRHO::DragonflyWrapperEditor::paint(juce::Graphics&) in lto.o
  "DISTRHO::PluginExporter::sFallbackString", referenced from:
      DISTRHO::DragonflyWrapperProcessor::getStateInformation(juce::MemoryBlock&) in lto.o
      DISTRHO::DragonflyWrapperProcessor::setStateInformation(void const*, int) in lto.o
      DISTRHO::ParameterFromDPF::getName(int) const in lto.o
      DISTRHO::ParameterFromDPF::getLabel() const in lto.o
  "DISTRHO::d_nextSampleRate", referenced from:
      JuceAU::JuceAU(ComponentInstanceRecord*) in lto.o
  "DISTRHO::createStaticPlugin()", referenced from:
      JuceAU::JuceAU(ComponentInstanceRecord*) in lto.o
  "DISTRHO::PluginExporter::sFallbackEnumValues", referenced from:
      JuceAU::JuceAU(ComponentInstanceRecord*) in lto.o
  "DISTRHO::PluginExporter::sFallbackRanges", referenced from:
      JuceAU::JuceAU(ComponentInstanceRecord*) in lto.o
      DISTRHO::DragonflyWrapperProcessor::setStateInformation(void const*, int) in lto.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I'm not sure what to do to get the arm64 part to work here. It's on a branch called `au-build` if you want to take a look.

**Comment by falkTX** [comment]
could be you only built for x64 while the juce cmake stuff is universal, or only linking to ui and not dsp.
that cmake juce thing is 100% plugin dependent and not generic

tbh you should be able to fix these issues otherwise it wont be a good idea to enable AU.
otherwise next time something changes and another build error happens you will be stuck again.

**Comment by falkTX** [comment]
this is coming soon :)

![image](https://github.com/michaelwillis/dragonfly-reverb/assets/1334853/a3a88672-32eb-44b8-b3ea-0b3e14898c9b)

**Comment by michaelwillis** [comment]
@falkTX I tried building the AU plugins: [https://github.com/michaelwillis/dragonfly-reverb/compare/master...au-build](https://github.com/michaelwillis/dragonfly-reverb/compare/master...au-build)

It mostly worked, I can hear the reverb working, but it doesn't render the background image and some of the widgets are the wrong scale. I got the same results in Ardour, Carla, and GarageBand. Do you have any suggestions?

<img width="1207" alt="Screenshot 2024-04-27 at 11 47 00 AM" src="https://github.com/michaelwillis/dragonfly-reverb/assets/740213/5bbe4021-dadf-4529-bb71-fa95a22b7b21">

**Comment by falkTX** [comment]
is that on macOS? with a high-dpi screen?

note the use of `DISTRHO_PLUGIN_BRAND_ID` breaks vst3 if using it for old/existing plugins as-is. see https://distrho.github.io/DPF/group__PluginMacros.html#ga9c9caf7e3376c9235c3529dbc612a964

**Comment by falkTX** [comment]
is this maybe a case of mixing a high-dpi screen (like laptop) with a regular external 100% scale one?

**Comment by michaelwillis** [comment]
Yes, I had a low-dpi monitor plugged into my MacBook. I'll try it again
without the monitor plugged in when I get back to that laptop.

**Comment by michaelwillis** [comment]
@falkTX I tried it again, I get the same graphics bugs on two different Macbooks without any external display connected.

* 2021 MacBook Pro M1, macOS Sonoma 14.4.1, 3456x2234 retina display 
* 2019 MacBook Pro, AMD Radeon Pro 5500M + Intel UHD 630, 3072x1920 retina display

**Comment by SomeoneCalledRob** [comment]
Any hope of this effort being resumed? I really dig the Dragonfly sound, but having recently gone from using an Intel to Apple Silicon Mac, I miss having a fully native build.

**Comment by falkTX** [comment]
we should try again with latest dpf develop branch, I recently pushed some changes to help prevent bad behaviour. just watch out for the deprecated calls and replace them with whats indicated in the deprecation notice.
